// @x402/evm/dist/esm/chunk-U4H6Q62Q.mjs
import {
  ExactEvmSchemeV1,
  NETWORKS,
  PERMIT2_ADDRESS,
  authorizationTypes,
  createNonce,
  createPermit2Nonce,
  permit2WitnessTypes,
  x402ExactPermit2ProxyAddress
} from "./chunk-DSSJHWGT.mjs";

// src/exact/client/eip3009.ts
import { getAddress } from "viem";
async function createEIP3009Payload(signer, x402Version, paymentRequirements) {
  const nonce = createNonce();
  const now = Math.floor(Date.now() / 1e3);
  const authorization = {
    from: signer.address,
    to: getAddress(paymentRequirements.payTo),
    value: paymentRequirements.amount,
    validAfter: (now - 600).toString(),
    validBefore: (now + paymentRequirements.maxTimeoutSeconds).toString(),
    nonce
  };
  const signature = await signEIP3009Authorization(signer, authorization, paymentRequirements);
  const payload = {
    authorization,
    signature
  };
  return {
    x402Version,
    payload
  };
}
async function signEIP3009Authorization(signer, authorization, requirements) {
  const chainId = parseInt(requirements.network.split(":")[1]);
  if (!requirements.extra?.name || !requirements.extra?.version) {
    throw new Error(
      `EIP-712 domain parameters (name, version) are required in payment requirements for asset ${requirements.asset}`
    );
  }
  const { name, version } = requirements.extra;
  const domain = {
    name,
    version,
    chainId,
    verifyingContract: getAddress(requirements.asset)
  };
  const message = {
    from: getAddress(authorization.from),
    to: getAddress(authorization.to),
    value: BigInt(authorization.value),
    validAfter: BigInt(authorization.validAfter),
    validBefore: BigInt(authorization.validBefore),
    nonce: authorization.nonce
  };
  return await signer.signTypedData({
    domain,
    types: authorizationTypes,
    primaryType: "TransferWithAuthorization",
    message
  });
}

// src/exact/client/permit2.ts
import { encodeFunctionData, getAddress as getAddress2 } from "viem";
var MAX_UINT256 = BigInt("0xffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff");
async function createPermit2Payload(signer, x402Version, paymentRequirements) {
  const now = Math.floor(Date.now() / 1e3);
  const nonce = createPermit2Nonce();
  const validAfter = (now - 600).toString();
  const deadline = (now + paymentRequirements.maxTimeoutSeconds).toString();
  const permit2Authorization = {
    from: signer.address,
    permitted: {
      token: getAddress2(paymentRequirements.asset),
      amount: paymentRequirements.amount
    },
    spender: x402ExactPermit2ProxyAddress,
    nonce,
    deadline,
    witness: {
      to: getAddress2(paymentRequirements.payTo),
      validAfter,
      extra: "0x"
    }
  };
  const signature = await signPermit2Authorization(
    signer,
    permit2Authorization,
    paymentRequirements
  );
  const payload = {
    signature,
    permit2Authorization
  };
  return {
    x402Version,
    payload
  };
}
async function signPermit2Authorization(signer, permit2Authorization, requirements) {
  const chainId = parseInt(requirements.network.split(":")[1]);
  const domain = {
    name: "Permit2",
    chainId,
    verifyingContract: PERMIT2_ADDRESS
  };
  const message = {
    permitted: {
      token: getAddress2(permit2Authorization.permitted.token),
      amount: BigInt(permit2Authorization.permitted.amount)
    },
    spender: getAddress2(permit2Authorization.spender),
    nonce: BigInt(permit2Authorization.nonce),
    deadline: BigInt(permit2Authorization.deadline),
    witness: {
      to: getAddress2(permit2Authorization.witness.to),
      validAfter: BigInt(permit2Authorization.witness.validAfter),
      extra: permit2Authorization.witness.extra
    }
  };
  return await signer.signTypedData({
    domain,
    types: permit2WitnessTypes,
    primaryType: "PermitWitnessTransferFrom",
    message
  });
}
var erc20ApproveAbi = [
  {
    type: "function",
    name: "approve",
    inputs: [
      { name: "spender", type: "address" },
      { name: "amount", type: "uint256" }
    ],
    outputs: [{ type: "bool" }],
    stateMutability: "nonpayable"
  }
];
var erc20AllowanceAbi = [
  {
    type: "function",
    name: "allowance",
    inputs: [
      { name: "owner", type: "address" },
      { name: "spender", type: "address" }
    ],
    outputs: [{ type: "uint256" }],
    stateMutability: "view"
  }
];
function createPermit2ApprovalTx(tokenAddress) {
  const data = encodeFunctionData({
    abi: erc20ApproveAbi,
    functionName: "approve",
    args: [PERMIT2_ADDRESS, MAX_UINT256]
  });
  return {
    to: getAddress2(tokenAddress),
    data
  };
}
function getPermit2AllowanceReadParams(params) {
  return {
    address: getAddress2(params.tokenAddress),
    abi: erc20AllowanceAbi,
    functionName: "allowance",
    args: [getAddress2(params.ownerAddress), PERMIT2_ADDRESS]
  };
}

// src/exact/client/scheme.ts
var ExactEvmScheme = class {
  /**
   * Creates a new ExactEvmClient instance.
   *
   * @param signer - The EVM signer for client operations
   */
  constructor(signer) {
    this.signer = signer;
    this.scheme = "exact";
  }
  /**
   * Creates a payment payload for the Exact scheme.
   * Routes to EIP-3009 or Permit2 based on requirements.extra.assetTransferMethod.
   *
   * @param x402Version - The x402 protocol version
   * @param paymentRequirements - The payment requirements
   * @returns Promise resolving to a payment payload result
   */
  async createPaymentPayload(x402Version, paymentRequirements) {
    const assetTransferMethod = paymentRequirements.extra?.assetTransferMethod ?? "eip3009";
    if (assetTransferMethod === "permit2") {
      return createPermit2Payload(this.signer, x402Version, paymentRequirements);
    }
    return createEIP3009Payload(this.signer, x402Version, paymentRequirements);
  }
};

// src/exact/client/register.ts
function registerExactEvmScheme(client, config) {
  if (config.networks && config.networks.length > 0) {
    config.networks.forEach((network) => {
      client.register(network, new ExactEvmScheme(config.signer));
    });
  } else {
    client.register("eip155:*", new ExactEvmScheme(config.signer));
  }
  NETWORKS.forEach((network) => {
    client.registerV1(network, new ExactEvmSchemeV1(config.signer));
  });
  if (config.policies) {
    config.policies.forEach((policy) => {
      client.registerPolicy(policy);
    });
  }
  return client;
}

export {
  erc20AllowanceAbi,
  createPermit2ApprovalTx,
  getPermit2AllowanceReadParams,
  ExactEvmScheme,
  registerExactEvmScheme
};
